extensions/README.md

```diff
@@ -0,0 +1,373 @@
+> they are defined in a different spec (and are not implemented in this
+> library), they cannot be extended using the above mechanisms.
+
+## Map Extensions
+
+Map extensions allow extending CoRIM maps with additional keys, effectively
+defining new fields for the corresponding structures. In the code base, these
+can be identified by the embedded `Extensions` struct. These are
+
+- `comid.Comid`
+- `comid.Entity`
+- `comid.FlagsMap`
+- `comid.Mval`
+- `comid.Triples`
+- `corim.Entity`
+- `corim.Signer`
+- `corim.UnsignedCorim`
+
+To extend the above types, you need to define a struct containing your
+extensions and pass a pointer to an instance of that struct to the
+`RegisterExtensions()` method of the corresponding instance of the type that is
+being extended. This should be done as early as possible, before any marshaling
+is performed.
+
+These types can be extended in two ways: by adding additional fields, and by
+introducing additional constraints over existing fields.
+
+### Adding new fields
+
+To add a new fields, simply add them to your extensions struct, ensuring that
+the `cbor` and `json` tags on those fields are set correctly. As CoRIM spec
+mandates integer keys, you must use `keyasint` option for the `cbor` tag.
+
+To access the values of those fields, you can call the extended type instance's
+`Extensions.Get()` passing in the name of the field you want to access. The
+name can be either the Go struct's field name, the name specified in the `json`
+tag, or (a string containing) the integer specified in the `cbor` tag.
+
+`Get()` returns an `interface{}`. There are equivalent `GetInt()`,
+`GetString()`, etc. methods that  perform the required conversions, and return
+the value of the indicated type, along with possible errors. ("Must" versions
+of these also exists, e.g. `MustGetString()`, that do not return an error).
+
+You can also get the pointer your extensions instance itself by calling
+extended type instance's `GetExtensions()`. This returns an `interface{}`, so
+you will need to type assert in order to be able to access the fields directly.
+
+### Introducing additional constraints
+
+To introduce new constraints,  add a method called `Validate<type>(v *<type>)`
+to your extensions struct, where `<type>` is the name of the type being
+extended (one of the ones listed above) -- e.g.
+`ValidateComid(v *comid.Comid)` when extending `comid.Comid`. This method, if
+it exists, will be invoked inside the extended type instance's `Valid()`
+method, passing itself as the parameter.
+
+You do not need to define this method, unless you actually want to enforce some
+constraints (i.e. if you just want to define additional fields).
+
+### Example
+
+The following example illustrates how to implement a map extension by extending `comid.Entity` with the following features:
+
+1. an optional "email" field
+2. additional validation to ensure that the existing name field contains a
```

Comment:
I have a clarification question here more than a comment.

What is the `use case` you are trying to highlight where the Validate method 
is checking for existing element validity ?

My concern is that a new user of extensions will get further confused on how the Validate is useful ?

For the benefit of simplicity can we keep it just checking the new parameters added for validity (like URI added is a valid URI?)

Reply:
You _can_ use Validate*  to  enforce individual field constraits on new fields, but it is not necessary for that, and it is not the reason it exists.  It is not necessary for that.

The things that the Valadiate* methods make possible that would not be possible other wise are:

- validating additional contraints on existing fields (since the types of those fields are fixed, you cannot do it the usual way, like you can with new fields)
- valdiating inter-field constraints (which, by their nature, cannot be captured in a type) 

The example illustrates that.

The primary mechanism for enforcing constraints on individual values is types. that's the sole reason for having types in the first place.

> OR I am setting an new Identifier which can be []byte but valid ONLY
when size is fixed to 32 bytes or 64 bytes.

You can do this by introducing your won type.  which is defined as "a seequence of exactly 32 or 64 bytes". 

If the example only showed validation of a new field, _that_ would be a sourc of confusion -- "why do I need to define Vadate method that take a whole other struct as an object, when I can just define my own type?"